src/Dapr.Messaging/PublishSubscribe/PublishSubscribeReceiver.cs

```diff
@@ -0,0 +1,305 @@
+    /// The stream connection between this instance and the Dapr sidecar.
+    /// </summary>
+    private AsyncDuplexStreamingCall<P.SubscribeTopicEventsRequestAlpha1, P.SubscribeTopicEventsResponseAlpha1>? clientStream;
+    /// <summary>
+    /// Used to ensure thread-safe operations against the stream.
+    /// </summary>
+    private readonly SemaphoreSlim semaphore = new(1, 1);
+    /// <summary>
+    /// The handler delegate responsible for processing the topic messages.
+    /// </summary>
+    private readonly TopicMessageHandler messageHandler;
+    /// <summary>
+    /// A reference to the DaprClient instance.
+    /// </summary>
+    private readonly P.Dapr.DaprClient client;
+    /// <summary>
+    /// Flag that prevents the developer from accidentally initializing the subscription more than once from the same receiver.
+    /// </summary>
+    private bool hasInitialized;
+    /// <summary>
+    /// Flag that ensures the instance is only disposed a single time.
+    /// </summary>
+    private bool isDisposed;
+
+    /// <summary>
+    /// Constructs a new instance of a <see cref="PublishSubscribeReceiver"/> instance.
+    /// </summary>
+    /// <param name="pubSubName">The name of the Dapr Publish/Subscribe component.</param>
+    /// <param name="topicName">The name of the topic to subscribe to.</param>
+    /// <param name="options">Options allowing the behavior of the receiver to be configured.</param>
+    /// <param name="handler">The delegate reflecting the action to take upon messages received by the subscription.</param>
+    /// <param name="client">A reference to the DaprClient instance.</param>
+    internal PublishSubscribeReceiver(string pubSubName, string topicName, DaprSubscriptionOptions options, TopicMessageHandler handler, P.Dapr.DaprClient client)
+    {
+        this.client = client;
+        this.pubSubName = pubSubName;
+        this.topicName = topicName;
+        this.options = options;
+        this.messageHandler = handler;
+        topicMessagesChannel = options.MaximumQueuedMessages is > 0
+            ? Channel.CreateBounded<TopicMessage>(new BoundedChannelOptions((int)options.MaximumQueuedMessages)
+            {
+                SingleReader = true, SingleWriter = true, FullMode = BoundedChannelFullMode.Wait
+            })
+            : Channel.CreateUnbounded<TopicMessage>(UnboundedChannelOptions);
+    }
+
+    /// <summary>
+    /// Dynamically subscribes to messages on a PubSub topic provided by the Dapr sidecar.
+    /// </summary>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>An <see cref="IAsyncEnumerable{TopicMessage}"/> containing messages provided by the sidecar.</returns>
+    internal async Task SubscribeAsync(CancellationToken cancellationToken = default)
+    {
+        //Prevents the receiver from performing the subscribe operation more than once (as the multiple initialization messages would cancel the stream).
+        if (hasInitialized)
```

Comment:
This doesn't completely eliminate the risk of a race (e.g. two threads in a thread pool).  That said, given this is now an internal method that is intended to (and we guarantee is) called only once, we might just dispense with it (and just document the method as "not thread safe").